### PR TITLE
Recommit "Implement [[msvc::no_unique_address]] (#65675)"

### PR DESCRIPTION
**Pull request by amykhuang**
This implements the [[msvc::no_unique_address]] attribute.

There is not ABI compatibility in this patch because the attribute is relatively new and there's still some uncertainty in the MSVC version.

The recommit changes the attribute definitions so that instead of making two separate attributes for no_unique_address
and msvc::no_unique_address, it modifies the attributes tablegen emitter to allow spellings to be target-specific. 

This reverts commit 71f9e7695b87298f9855d8890f0e6a3b89381eb5.

### CODE REVIEW

**Comment by amykhuang** [comment]
Previous commit attempt caused sphinx errors. Also, because we now have two attributes for the itanium and MSVC versions, there are two separate sections in the documentation, which I didn't realize before.

File: clang/include/clang/Basic/Attr.td [hunk 1]
```diff
@@ -1798,11 +1798,24 @@ def ArmMveStrictPolymorphism : TypeAttr, TargetSpecificAttr<TargetARM> {
...
+  let Documentation = [InternalOnly];
+}
+
+def MSNoUniqueAddress : InheritableAttr, TargetSpecificAttr<TargetMicrosoftCXXABI> {
   let Subjects = SubjectList<[NonBitField], ErrorDiag>;
+  let Spellings = [CXX11<"msvc", "no_unique_address", 201803>];
+  let Documentation = [MSNoUniqueAddressDocs];
+}
+
+def ItaniumNoUniqueAddress : InheritableAttr, TargetSpecificAttr<TargetItaniumCXXABI> {
```
Comment (zygoloid) [inline clang/include/clang/Basic/Attr.td:1815]: I don't think this attribute name makes sense: this is the standard attribute (which just happens to have no effect on the MS ABI), not something specific to the Itanium ABI.

Comment (amykhuang) [reply clang/include/clang/Basic/Attr.td:1815]: Yeah, that makes sense; would it be better to name it "NoUniqueAddress" and then name the "parent" attribute something like "NoUniqueAddressGroup"?

Comment (zygoloid) [reply clang/include/clang/Basic/Attr.td:1815]: Perhaps the nicest thing would be to change the tablegen handler to support marking individual spellings as target-specific, if that's easy to do.

**Review by zygoloid — APPROVED** [review]
Looks awesome, thanks :)

**Comment by zygoloid** [comment]
(Please remember to update the commit message to describe the tablegen change when merging.)

**Comment by amykhuang** [comment]
hmm this passes most of the checks in SemaCXX/cxx2a-no-unique-address.cpp, but fails "attribute 'no_unique_address' cannot have an argument list". Also fails Preprocessor/has_attribute.cpp because both attributes are recognized on itanium and windows targets.

**Comment by amykhuang** [comment]
Added a section for target specific spellings in `GenerateHasAttrSpellingStringSwitch`, and a check for "existsInTarget" in `ParseCXX11AttributeArgs`, and looks like all tests are passing now.

**Comment by cjdb** [comment]
Thanks @amykhuang! 🎉

**Pull request merged**